internal/querynodev2/delegator/delegator_data.go

```diff
@@ -0,0 +1,440 @@
+		growing := sd.segmentManager.GetGrowing(segmentID)
+		if growing == nil {
+			growing = sd.newGrowing(segmentID, insertData)
+		}
+
+		err := growing.Insert(insertData.RowIDs, insertData.Timestamps, insertData.InsertRecord)
+		if err != nil {
+			log.Error("failed to insert data into growing segment",
+				zap.Int64("segmentID", segmentID),
+				zap.Error(err),
+			)
+			// panic here, insert failure
+			panic(err)
+		}
+		growing.UpdateBloomFilter(insertData.PrimaryKeys)
+
+		log.Debug("insert into growing segment",
+			zap.Int64("collectionID", growing.Collection()),
+			zap.Int64("segmentID", segmentID),
+			zap.Int("rowCount", len(insertData.RowIDs)),
+			zap.Uint64s("tss", insertData.Timestamps),
+		)
+	}
+}
+
+// ProcessDelete handles delete data in delegator.
+// delegator puts deleteData into buffer first,
+// then dispatch data to segments acoording to the result of pkOracle.
+func (sd *shardDelegator) ProcessDelete(deleteData []*DeleteData, ts uint64) {
+	log := sd.getLogger(context.Background())
+
+	log.Debug("start to process delete", zap.Uint64("ts", ts))
+	// add deleteData into buffer.
+	cacheItems := make([]deleteBufferItem, 0, len(deleteData))
+	for _, entry := range deleteData {
+		cacheItems = append(cacheItems, deleteBufferItem{
+			partitionID: entry.PartitionID,
+			deleteData: storage.DeleteData{
+				Pks:      entry.PrimaryKeys,
+				Tss:      entry.Timestamps,
+				RowCount: entry.RowCount,
+			},
+		})
+	}
+	sd.deleteBuffer.Cache(ts, cacheItems)
+
+	// segment => delete data
+	delRecords := make(map[int64]DeleteData)
+	for _, data := range deleteData {
+		for i, pk := range data.PrimaryKeys {
+			segmentIDs, err := sd.pkOracle.Get(pk, pkoracle.WithPartitionID(data.PartitionID))
+			if err != nil {
+				log.Warn("failed to get delete candidates for pk", zap.Any("pk", pk.GetValue()))
+				continue
+			}
+			for _, segmentID := range segmentIDs {
+				delRecord := delRecords[segmentID]
+				delRecord.PrimaryKeys = append(delRecord.PrimaryKeys, pk)
+				delRecord.Timestamps = append(delRecord.Timestamps, data.Timestamps[i])
+				delRecords[segmentID] = delRecord
+			}
+		}
+	}
+
+	applyDelete := func(nodeID int64, worker cluster.Worker, entries []SegmentEntry) {
+		for _, segmentEntry := range entries {
+			delRecord, ok := delRecords[segmentEntry.SegmentID]
+			if ok {
+				log.Debug("delegator plan to applyDelete via worker", zap.Int64("nodeID", nodeID), zap.Int64("segmentID", segmentEntry.SegmentID))
+				err := worker.Delete(context.Background(), &querypb.DeleteRequest{
+					Base:         commonpbutil.NewMsgBase(commonpbutil.WithTargetID(nodeID)),
+					CollectionId: sd.collectionID,
+					PartitionId:  segmentEntry.PartitionID,
+					VchannelName: sd.vchannelName,
+					SegmentId:    segmentEntry.SegmentID,
+					PrimaryKeys:  storage.ParsePrimaryKeys2IDs(delRecord.PrimaryKeys),
+					Timestamps:   delRecord.Timestamps,
+				})
+				if err != nil {
+					log.Warn("worker failed to delete on segment",
```

Comment:
can we just ignore the error if delete applied failed?

if we skip here but the segment owner is alive, then the tsafe is pushed forward and we may search out deleted data